### PR TITLE
Support specifying a custom subnet for ILB ip in GCE

### LINKED ISSUE #81966: GCE InternalLoadBalancer does not accept a custom subnet for IP address allocation
**What happened**:
ILB ip addresses come from the subnet that the K8s cluster uses. There is no ability to specify a different subnet.
**What you expected to happen**:
ILB ip addresses should be allocated from a different subnet if specified. 
**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration: GCE
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by prameshj**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR allows GCE users to specify a custom subnetwork to assign Internal LoadBalancer IP from 
**Which issue(s) this PR fixes**:

Fixes #81966

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by prameshj** [comment]
Since 1.16 is in code freeze, we will target this change for 1.17.

/aasign @bowei for review

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go [hunk 1]
```diff
@@ -103,13 +103,26 @@ func (g *Cloud) ensureInternalLoadBalancer(clusterName, clusterID string, svc *v
...
+		// Removing the annotation will set the forwarding rule to use the default subnet.
+		if options.SubnetName != "" {
+			if existingFwdRule != nil && options.SubnetName != existingFwdRule.Subnetwork {
+				// Subnet has changed, reset ipToUse to IP provided in the spec, if any.
+				ipToUse = svc.Spec.LoadBalancerIP
+			}
+			subnetworkURL = gceSubnetworkURL("", g.networkProjectID, g.region, options.SubnetName)
+		}
+	} else {
+		if existingFwdRule != nil && existingFwdRule.Subnetwork != "" {
```
Comment (bowei) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go]: Is this what we want? This seems to violate k8s-style reconciliation

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go [hunk 2]
```diff
@@ -736,7 +749,9 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
 }
 
 func getILBOptions(svc *v1.Service) ILBOptions {
-	return ILBOptions{AllowGlobalAccess: GetLoadBalancerAnnotationAllowGlobalAccess(svc)}
+	return ILBOptions{AllowGlobalAccess: GetLoadBalancerAnnotationAllowGlobalAccess(svc),
```
Comment (bowei) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go]: don't eat the newline here:
```
ILBOptions{
  AllowGlobalAccess: ..., 
  ...,
}
```

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go (hunk 1 above)
Comment (bowei) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go]: I see this was logic from before, which seems kind of strange.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go [hunk 3]
```diff
@@ -103,13 +103,26 @@ func (g *Cloud) ensureInternalLoadBalancer(clusterName, clusterID string, svc *v
...
 	subnetworkURL := g.SubnetworkURL()
-	if existingFwdRule != nil && existingFwdRule.Subnetwork != "" {
-		// external LBs have an empty Subnetwork field.
-		subnetworkURL = existingFwdRule.Subnetwork
+	if g.AlphaFeatureGate.Enabled(AlphaFeatureILBCustomSubnet) {
+		// If this feature is enabled, changes to subnet annotation will be
+		// picked up and reflected in the forwarding rule.
+		// Removing the annotation will set the forwarding rule to use the default subnet.
+		if options.SubnetName != "" {
+			if existingFwdRule != nil && options.SubnetName != existingFwdRule.Subnetwork {
```
Comment (bowei) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go]: I might be missing something -- why does the LoadBalancerIP get reused? When you change subnets, they will have different IP ranges.

Comment (prameshj) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go]: Correct, the code here only sets ipToUse to an IP specified by the user. This is to support the case where the user picks both the subnet and a specific IP. If the selected IP is not part of the subnet, the forwarding rule creation will error out.
svc.Spec.LoadBalancerIP will be empty in other cases, so we will be reseting the IP to use, if subnet changed.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go [hunk 4]
```diff
@@ -723,20 +738,27 @@ func getNameFromLink(link string) string {
 	return fields[len(fields)-1]
 }
 
-func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) string {
+func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule, requestedSubnet string) string {
```
Comment (bowei) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go]: // ilbIPToUse

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go [hunk 5]
```diff
@@ -723,20 +738,27 @@ func getNameFromLink(link string) string {
...
 		return svc.Spec.LoadBalancerIP
 	}
+	var ipToUse string
 
 	if fwdRule != nil {
-		return fwdRule.IPAddress
+		ipToUse = fwdRule.IPAddress
+		if requestedSubnet != fwdRule.Subnetwork {
+			// reset ip address since subnet is being changed.
+			ipToUse = ""
```
Comment (bowei) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go]: just return directly here

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go [hunk 6]
```diff
@@ -758,20 +773,30 @@ func getNameFromLink(link string) string {
...
+// that is reused. In case a subnetwork change is requested, the existing ForwardingRule IP is ignored.
+func ilbIPToUse(svc *v1.Service, fwdRule *compute.ForwardingRule, requestedSubnet string) string {
 	if svc.Spec.LoadBalancerIP != "" {
 		return svc.Spec.LoadBalancerIP
 	}
+	var ipToUse string
 
 	if fwdRule != nil {
-		return fwdRule.IPAddress
+		ipToUse = fwdRule.IPAddress
```
Comment (bowei) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go]: logic can be simpler

```golang
if svc.Spec.LoadBalancerIP != "" {
  return svc.Spec.LoadBalancerIP
}
if fwdRule == nil {
  return ""
}
if requestedSubnet != fwdRule.Subnetwork {
  return ""
}
return fwdRule.IPAddress
```

Comment (prameshj) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go]: done

**Pull request merged**